add scmp and ucmp to SelectionDAG.cpp

miguelraz opened the pull request:
This is a PR based on https://github.com/llvm/llvm-project/pull/83227 that tries to add the `scmp` and `ucmp` new intrinsics to `SelectionDAG.cpp` .

cc @nikic

miguelraz committed:
add sthreecmp and uthreecmp intrinsics to LangRef

miguelraz committed:
add tablegen for sthreecmp and usthreecmp intrinsics

miguelraz committed:
typo fix unsigned integer for uthreecmp

labrinea committed:
[clang] Remove unused lambda capture. (#83550)

Fixes the `sanitizer-x86_64-linux-android` buildbot.

miguelraz committed:
add sthreecmp and uthreecmp intrinsics modeled after llvm.smax

miguelraz committed:
add int_sthreecmp and int_uthreecmp based on int_smin

miguelraz committed:
add [u]sthreecmp to ValueTrackingTest.cpp

AZero13 committed:
[Hexagon] Use LiveRegUnits (#84112)

miguelraz committed:
call three way cmp intrinsic [u]scmp in LangRef

miguelraz committed:
only add minimum stuff to LangRef and Intrinsics.td for s/ucmp

miguelraz committed:
first changes for ValueTrackingTest.cpp and SelectionDAG for s/ucmp

miguelraz committed:
add first attempt at SelectionDAG.cpp for u/scmp

miguelraz committed:
fix formatting on irrelevant lines

topperc commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -6747,12 +6747,18 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
-    assert(VT.isFloatingPoint() && "This operator only applies to FP types!");
+  case ISD::UCMP:
+  case ISD::SCMP:
+    assert(VT.isInteger() && "This operator doe snot apply to FP types!");
```
"doe snot" -> "does not"

miguelraz replied:
Ah, good catch!

miguelraz committed:
fix typo in assert

miguelraz committed:
add u/scmp result type overload over a fixed type in Intrinsics.td

miguelraz committed:
add VisitIntrinsicCall cases for Intrinsic::ucmp/scmp

miguelraz committed:
Update llvm/lib/IR/Verifier.cpp

Co-authored-by: Nikita Popov <github@npopov.com>

miguelraz committed:
add proper type overloads for llvm.ucmp.i2.i32 and llvm.scmp.i2.i32 in LangRef

miguelraz committed:
move isVector check above else to catch dest as scalar but src as vector

miguelraz committed:
omit checks done by Intrinsics.td on u/scmp operating on int-likes

miguelraz committed:
Update llvm/docs/LangRef.rst

Co-authored-by: Nikita Popov <github@npopov.com>

miguelraz committed:
Update llvm/docs/LangRef.rst

Co-authored-by: Nikita Popov <github@npopov.com>

miguelraz committed:
Update llvm/docs/LangRef.rst

Co-authored-by: Nikita Popov <github@npopov.com>

miguelraz committed:
Update llvm/lib/IR/Verifier.cpp

Co-authored-by: Nikita Popov <github@npopov.com>

miguelraz committed:
Update llvm/lib/IR/Verifier.cpp

Co-authored-by: Nikita Popov <github@npopov.com>

miguelraz committed:
delete stray newline

miguelraz committed:
cast to vectors (not arrays!) in scmp Check

miguelraz committed:
add Filecheck tests for Verifier.cpp u/scmp intrinsics

miguelraz committed:
delete last unneded CHECK

miguelraz committed:
make return value length more simple on scmp and ucmp

miguelraz committed:
make return value length more simple on scmp and ucmp

miguelraz committed:
Update llvm/docs/LangRef.rst

Co-authored-by: scottmcm <scottmcm@users.noreply.github.com>

miguelraz committed:
Update llvm/docs/LangRef.rst

Co-authored-by: scottmcm <scottmcm@users.noreply.github.com>

amykhuang committed:
Unrevert "[Clang][C++23] Implement P2448R2: Relaxing some constexpr re… (#85140)

Unrevert commit
https://github.com/llvm/llvm-project/commit/2f67dfb012038678dd0b873394a55bd5c937f843
as there were already dependent changes in the codebase that now fail.

miguelraz committed:
Update llvm/docs/LangRef.rst

Co-authored-by: Nikita Popov <github@npopov.com>

miguelraz committed:
Update llvm/docs/LangRef.rst

Co-authored-by: Nikita Popov <github@npopov.com>

miguelraz committed:
Update llvm/lib/IR/Verifier.cpp

Co-authored-by: Nikita Popov <github@npopov.com>

miguelraz committed:
fix CodingStandards for var casenames and 4 spaces not tabs in FileCheck tests

miguelraz committed:
Update llvm/test/Verifier/intrinsic-cmp.ll

Co-authored-by: Nikita Popov <github@npopov.com>

miguelraz committed:
align to 2 spaces

miguelraz committed:
fixup unit tests for scmp intrinsic

miguelraz committed:
appease the formatting gods

miguelraz committed:
appease the formatting gods, again

miguelraz committed:
appease the formatting gods, again, again

miguelraz committed:
appease the formatting gods, again, again, again

miguelraz committed:
appease the FileCheck gods \o/

miguelraz committed:
phrasing nits for error messages

miguelraz committed:
update FileCheck unit tests with most recent errors

miguelraz committed:
freakin' formatting gods

miguelraz commented:
Closing due to too many git headaches and restarting on fresh main with https://github.com/llvm/llvm-project/pull/85822
